[WIP] Add a generic reference data comparator

aferrero2707 opened the pull request:
The ReferenceComparator post-processing task allows to compare the plots and graphs produced by given task(s) with that of a reference run, and asses their compatibility. The syntax of the configuration allows to group the plots, for example by detector or by input task. The results are summarized in a canvas similar to that of the [BigScreen](https://github.com/AliceO2Group/QualityControl/blob/master/doc/PostProcessing.md#the-bigscreen-class) task.

The comparison can be based on either a chi2 probability criteria, or a maximum average deviation between actual and reference plots. The task can compare either histograms (1D or 2D) or graphs.

Barthelemy commented:
Hi @aferrero2707 , 

Thanks a lot, great that you started working on that.

Very sorry for the long delay. We need to rediscuss among us before coming back to you. Indeed we should make sure that it articulates well with https://its.cern.ch/jira/browse/QC-885 and https://its.cern.ch/jira/browse/QC-934 that were planned for after YETS. 

We shouldn't be long. 
Cheers

aferrero2707 commented:
@Barthelemy @knopers8 I have updated the code of the reference plot comparator, replacing the checker by a post-processing task such that we can easily process data from multiple detectors.
The output of the task is practically the same, a BigScreen-like matrix of colored boxes showing the overall result of the comparison for the different detectors.
The task also creates individual canvases for each group. Each canvas lists the plots in the group by name, and the text is colored according to the result of the comparison. If needed, this allows to easily identify the plots that deviate too much from the reference, like this:

<img width="1480" alt="image" src="https://github.com/AliceO2Group/QualityControl/assets/6229237/0f328522-4781-4d2e-921d-cce4536b478c">


The syntax of the configuration allows to create arbitrary groups of plots. So the task could also be used by a single detector, to compare plots with reference and groups them logically (for example check separately the output of digits, clusters, tracks, PID, etc...).

Another use of the module could be the validation of the output of a new software release, by comparing it with the last good one.

The source of reference plots is currently specified via a run number in the JSON configuration, but the code can be easily adapted once the retrieval of reference plots is implemented in the framework.

I have also added some documentation with the JSON syntax.

For me this is ready for review.

aferrero2707 commented:
I have done some more work on the comparison between histograms and references. The new code provides a generic post-processing task (`ReferenceComparatorTask`) that allows to draw superimposed a given histogram and its reference, and also draw their ratio.
It also provides a generic checker (`ReferenceComparatorCheck`) that assesses the compatibility with the reference based on the content of the ratio plot.
The task with the BigScreen-like output described in the previous comment is also still there, and called `ReferenceValidatorTask`.

Barthelemy commented on Modules/Common/src/ReferenceValidatorTask.cxx:
```diff
@@ -0,0 +1,265 @@
+namespace o2::quality_control_modules::common
+{
+
+static bool splitObjectPath(std::string fullPath, std::string& path, std::string& name)
```
This function already exist in another class in this PR.

Barthelemy commented on Modules/Common/src/ReferenceValidatorTask.cxx:
```diff
@@ -0,0 +1,265 @@
+// The second element of the pair is set to true if the MO has a time stamp more recent than a user-supplied threshold
+
+static std::pair<std::shared_ptr<MonitorObject>, bool> getMO(repository::DatabaseInterface& qcdb, std::string fullPath, Trigger t, long notOlderThan)
+{
```
this method already exists somewhere else in this PR if I am not mistaken.

Barthelemy commented on Modules/Common/src/ReferenceValidatorTask.cxx:
```diff
@@ -0,0 +1,265 @@
+//
+// Compate a MonitorObject with the corresponding reference, using the appropriate checker method
+// based on the actual object inheritance (TH1, TGraph or TCanvas)
+
```
idem

Barthelemy commented on Modules/Common/src/ReferenceValidatorTask.cxx:
```diff
@@ -0,0 +1,265 @@
+
+template <>
+std::string ReferenceValidatorTask::getParameter(std::string parName, const std::string defaultValue, const o2::quality_control::core::Activity& activity)
+{
```
idem

Barthelemy commented on Modules/Common/src/ReferenceValidatorTask.cxx:
```diff
@@ -0,0 +1,265 @@
+
+template <>
+std::string ReferenceValidatorTask::getParameter(std::string parName, const std::string defaultValue)
+{
```
idem

Barthelemy reviewed (changes requested):
Hi, 

Thanks a lot for all this work, and sorry for the delay. 

This is quite a large PR. Perhaps next time you could try to split it into smaller pieces (but still consistent pieces). 

I agree with what is proposed. 

General comments: 
- Naming convention: the naming does not always follow the coding guidelines (avoid abbreviation, upper/lower case, have precise names). I marked a few but not all. Of course this should be used as guidelines and it does not always make sense. 
- Some functions are duplicated several times. I marked some of them but I might have missed others. 

I have put a number of comments that I let you check.

Barthelemy commented on Modules/Common/src/BigScreenCanvas.cxx:
```diff
@@ -0,0 +1,114 @@
+  float yl = y2 + 0.01;
+
+  // add pave associated to the detector
+  mPaves[name] = std::make_unique<TPaveText>(x1, y1, x2, y2);
```
Yes correct, I was not sure that it was what you wanted though. But apparently it is :-D

aferrero2707 commented on Modules/Common/src/BigScreenCanvas.cxx:
```diff
@@ -0,0 +1,114 @@
+{
+  float dx = 1.0 / mNCols;
+  float dy = 1.0 / mNRows;
+  float paveHalfWidth = 0.3 * dx;
```
This is to allow some padding (empty space) between the boxes. I will make it more clear and explicit in the code.

aferrero2707 commented on Modules/Common/CMakeLists.txt:
```diff
@@ -47,7 +54,11 @@ add_root_dictionary(O2QcCommon
                             include/Common/BigScreen.h
+                            include/Common/ReferenceComparatorTask.h
+                            include/Common/ReferenceComparatorCheck.h
+                            include/Common/ReferenceValidatorTask.h
```
If I am not wrong, the missing headers are:
```
BigScreenCanvas.h
ReferenceComparatorAlgo.h
ReferenceComparatorTaskConfig.h
ReferenceValidatorTaskConfig.h
```

My impression is that this kind of headers, also from other parts of the code, are not listed in the root dictionary.
I only added the headers directly containing the objects that are loaded at runtime.
But I might be missing something here.

aferrero2707 commented on Modules/Common/src/BigScreenCanvas.cxx:
```diff
@@ -0,0 +1,114 @@
+
+void BigScreenCanvas::add(std::string name, int index)
+{
+  float dx = 1.0 / mNCols;
```
Indeed, TPad::Divide might simplify the code. I will give it a try and see if it becomes more readable.

aferrero2707 replied:
Already done in https://github.com/AliceO2Group/QualityControl/pull/2192

aferrero2707 commented on Modules/Common/src/BigScreenCanvas.cxx:
```diff
@@ -0,0 +1,114 @@
+
+void BigScreenCanvas::set(std::string name, int color, std::string text)
+{
+  auto pave = mPaves[name].get();
```
Indeed you are right, here it should be protected. It will not crash if `name` is not initialize, but it will add a null element, which should be better avoided.

aferrero2707 replied:
Fixed in https://github.com/AliceO2Group/QualityControl/pull/2192

aferrero2707 commented on Modules/Common/src/ReferenceComparatorTask.cxx:
```diff
@@ -0,0 +1,523 @@
+  uint64_t timeStamp = 0;
+  activity.mId = run;
+  const auto filterMetadata = activity_helpers::asDatabaseMetadata(activity, false);
+  const auto objectValidity = qcdb->getLatestObjectValidity(activity.mProvenance + "/" + fullPath, filterMetadata);
```
@Barthelemy indeed! The call in the `getMOFromRun()` function only happens once at the start of the run, when the reference plot is searched and retrieved.
However, the code contains a similar call in `getMO()`, which is invoked at each trigger. Currently I could not find a more efficient way of retrieving the last valid object and checking its validity range in one single CCDB access, but this might well be due to my limited knowledge of the CCDB interface.
The issue here is that we cannot rely on the time stamp of the trigger to lookup the CCDB objects, in the case we use a timer-based trigger, since the validity of good objects might have expired shortly before the trigger is invoked.

aferrero2707 commented on Modules/Common/include/Common/ReferenceComparatorTask.h:
```diff
@@ -0,0 +1,123 @@
+template <class T>
+T ReferenceComparatorTask::getParameter(std::string parName, const T defaultValue, const o2::quality_control::core::Activity& activity)
+{
+  T result = defaultValue;
```
@Barthelemy would you consider adding this kind of template method directly into the `CustomParameters` class? I think it could be useful and cover a lot of use cases.

aferrero2707 commented on Modules/Common/src/ReferenceComparatorCheck.cxx:
```diff
@@ -0,0 +1,178 @@
+{
+
+template <>
+std::string ReferenceComparatorCheck::getParameter(std::string parName, const std::string defaultValue, const o2::quality_control::core::Activity& activity)
```
See here: https://github.com/AliceMCH/QualityControl/commit/e8039232e0c6ea1ecfa89ad6762b837dfbe1b7f9

aferrero2707 commented on Modules/Common/src/ReferenceComparatorCheck.cxx:
```diff
@@ -0,0 +1,178 @@
+  if (compMethodStr == "Chi2") {
+    mPlotCompMethod = MOCompChi2;
+  }
+  mCompThreshold = getParameter<double>("threshold", 0.1, activity);
```
This is now replaced by a loadable comparator object

aferrero2707 commented on Modules/Common/include/Common/ReferenceComparatorAlgo.h:
```diff
@@ -0,0 +1,35 @@
+namespace o2::quality_control_modules::common
+{
+
+enum MOCompMethod {
```
I have added a loadable comparator object, as you suggested. It is indeed a good point.

aferrero2707 commented:
@Barthelemy I have pushed a new version that should address most of your comments:
* The new code consists of a [post-processing task](https://github.com/AliceMCH/QualityControl/blob/reference-data-checker/Modules/Common/include/Common/ReferenceComparatorTask.h) that generates the comparison plots - actual/reference drawn superimposed, and ratio between the two - and a [checker](https://github.com/AliceMCH/QualityControl/blob/reference-data-checker/Modules/Common/include/Common/ReferenceComparatorCheck.h) that decides about the compatibility with the reference, using the ratio plots as inputs
* I introduced an interface for dynamically loadable plot comparators, with a [basic example](https://github.com/AliceMCH/QualityControl/blob/reference-data-checker/Modules/Common/include/Common/ObjectComparatorDeviation.h) that compares plots based on the average difference between bins - it should be ok for most of the uses cases
* the [checker](https://github.com/AliceMCH/QualityControl/blob/reference-data-checker/Modules/Common/include/Common/ReferenceComparatorCheck.h) now also sets quality reasons, that can be displayed in the QualitySummary
* I have tried to simplify the code, avoid code duplication, add more inline comments, and better naming of variables (but I might have missed some)

Here is an example of output from the post-processing task and the checker:
<img width="1266" alt="Screenshot 2024-04-20 at 16 56 02" src="https://github.com/AliceO2Group/QualityControl/assets/6229237/d6ebd9ae-bdf9-40ac-844e-45fb4e6c4f29">
